lib/publiccloud/basetest.pm

```diff
@@ -107,39 +107,64 @@ sub finalize {
+    diag($text . '$flags->{fatal}=' . $flags->{fatal}) if ($flags->{fatal});
+    diag($text . '$self->{result}=' . $self->{result}) if ($self->{result});
 
     if ($self->{run_args}) {
-        if ($self->{run_args}->{my_instance}) {
-            my $dumpable_instance = Storable::dclone($self->{run_args}->{my_instance});
-            $dumpable_instance->{provider}->{provider_client}->{credentials_file_content} = '******';
-            diag('Public Cloud finalize: $self->{run_args}->{my_instance}=' . Dumper($dumpable_instance));
-        }
+        diag($text . '$self->{run_args}= ' . $self->{run_args});
+        # check Provider
         if ($self->{run_args}->{my_provider}) {
             my $dumpable_provider = Storable::dclone($self->{run_args}->{my_provider});
             $dumpable_provider->{provider_client}->{credentials_file_content} = '******';
-            diag('Public Cloud finalize: $self->{run_args}->{my_provider}=' . Dumper($dumpable_provider));
+            diag($text . '$self->{run_args}->{my_provider}= ' . Dumper($dumpable_provider));
+        } else {
+            $text .= 'Provider is undefined or incomplete (ref.: $self->{run_args}->{my_provider})';
+            record_info('UNDEF. provider', $text);
+            diag($text);
+            # nothing to do, exiting:
+            return;
         }
+        # check Instance
+        if ($self->{run_args}->{my_instance}) {
+            my $dumpable_instance = Storable::dclone($self->{run_args}->{my_instance});
+            $dumpable_instance->{provider}->{provider_client}->{credentials_file_content} = '******';
+            $text .= '$self->{run_args}->{my_instance}=' . Dumper($dumpable_instance);
+            $upload = 1;
+        } else {
+            $text .= 'Instance is undefined or incomplete (ref.: $self->{run_args}->{my_instance})';
+            record_info('UNDEF. instance', $text);
+        }
+        diag($text);
+    } else {
+        $text .= 'running arguments undefined (ref.: $self->{run_args}).';
+        record_info('FAILED finalize', $text);
+        diag($text . 'Probably early errors.');
+        # nothing to do, exiting:
+        return;
     }
-
+    $text = 'Public Cloud finalize: ';
     # currently we have two cases when teardown of instance will be skipped:
     # 1. Job should have 'PUBLIC_CLOUD_NO_TEARDOWN' variable
     if (get_var('PUBLIC_CLOUD_NO_TEARDOWN')) {
-        diag('Public Cloud finalize: The test has PUBLIC_CLOUD_NO_TEARDOWN variable.');
-        eval { $self->_upload_logs() } or record_info('FAILED', "\$self->_upload_logs() failed -- $@", result => 'fail');
+        diag($text . 'The test has PUBLIC_CLOUD_NO_TEARDOWN variable.');
+        if ($upload) {
+            eval { $self->_upload_logs() }
+              or record_info('FAILED upload_logs', $text . "Failed _upload_logs no teardown (ref.: \$self->_upload_logs) --\n $@");
```

Comment:
```suggestion
              or record_info('FAILED', $text . "Failed _upload_logs no teardown (ref.: \$self->_upload_logs) --\n $@");
```

no more than one word

Reply:
ok , I was not aware ( or it was implemented only recently ) that hover showing you full text